### PR TITLE
Count on the controller side to pass only paginated set to the view

### PR DESCRIPTION
**Pull request by ncounter**
Compute the counting on the controller side, so the `@notifications` subset passed to the view is only the paginated visible one, and this deduplicate the execution of the query that occurs twice at the moment because of the `@all_filtered_notifications = @notifications`

### CODE REVIEW

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 1]
```diff
@@ -67,6 +68,24 @@ def set_notifications
...
+    @counted_notifications['comments'] = @notifications.for_comments.count
+    @counted_notifications['requests'] = @notifications.for_requests.count
+    @counted_notifications['incoming_requests'] = @notifications.for_incoming_requests(User.session).count
+    @counted_notifications['outgoing_requests'] = @notifications.for_outgoing_requests(User.session).count
+    @counted_notifications['relationships_created'] = @notifications.for_relationships_created.count
+    @counted_notifications['relationships_deleted'] = @notifications.for_relationships_deleted.count
+    @counted_notifications['build_failures'] = @notifications.for_build_failures.count
+    @counted_notifications['reports'] = @notifications.for_reports.count
+    @counted_notifications['workflow_runs'] = @notifications.for_workflow_runs.count
+    @counted_notifications['appealed_decisions'] = @notifications.for_appealed_decisions.count
```
Comment (eduardoj) [inline src/api/app/controllers/webui/users/notifications_controller.rb:85]:
```suggestion
```
Shouldn't adding all the filters back have a negative inpact on the users with lots of notifications?

I would prefer to introduce all these filters gradually. I don't think this is a good idea.

Comment (ncounter) [reply src/api/app/controllers/webui/users/notifications_controller.rb]: Deal https://github.com/openSUSE/open-build-service/pull/16260

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml [hunk 2]
```diff
@@ -1,102 +1,105 @@
 -# haml-lint:disable ViewLength
+- projects_for_filter = ProjectsForFilterFinder.new.call
+- groups_for_filter = GroupsForFilterFinder.new.call
```
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/_notification_filter.html.haml:3]: Those two are queries too...

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml (hunk 2 above)
Comment (ncounter) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml]: True. Moved to the controller too

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml [hunk 3]
```diff
@@ -4,99 +4,99 @@
...
       .px-4.py-1
-        = render partial: 'webui/shared/check_box', locals: { label: 'Appealed Decisions', amount: @count['appealed_decisions'],
+        = render partial: 'webui/shared/check_box', locals: { label: 'Appealed Decisions', amount: counted_notifications['appealed_decisions'],
                                                               key: 'kind[appealed_decisions]', name: 'kind[]', value: 'appealed_decisions',
                                                               label_icon: 'fas fa-hand',
-                                                              checked: @selected_filter[:kind]&.include?('appealed_decisions') }
+                                                              checked: selected_filter[:kind]&.include?('appealed_decisions') }
 
-  - unless @projects_for_filter.empty?
+  - unless projects_for_filter.empty?
```
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/_notification_filter.html.haml:74]: This is `@projects_for_filter` again

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml [hunk 4]
```diff
@@ -4,99 +4,99 @@
...
+                                                            checked: selected_filter[:kind]&.include?('relationships_deleted') }
     .px-4.py-1
-      = render partial: 'webui/shared/check_box', locals: { label: 'Build Failures', amount: @count['build_failures'],
+      = render partial: 'webui/shared/check_box', locals: { label: 'Build Failures', amount: counted_notifications['build_failures'],
                                                             key: 'kind[build_failures]', name: 'kind[]', value: 'build_failures',
                                                             label_icon: 'fas fa-xmark text-danger',
-                                                            checked: @selected_filter[:kind]&.include?('build_failures') }
-    - if ReportPolicy.new(@user, Report).notify?
+                                                            checked: selected_filter[:kind]&.include?('build_failures') }
+    - if ReportPolicy.new(user, Report).notify?
```
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/_notification_filter.html.haml:56]:
```suggestion
    - if policy(Report.new).notify?
```

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml (hunk 3 above)
Comment (ncounter) [reply src/api/app/views/webui/users/notifications/_notification_filter.html.haml:74]: Not really, it is passed as a local parameter in the `render partial` call
```
= render partial: 'notification_filter', locals: { selected_filter: @selected_filter,
                                                           counted_notifications: @counted_notifications,
                                                           user: User.session,
                                                           projects_for_filter: @projects_for_filter,
                                                           groups_for_filter: @groups_for_filter }
```

File: src/api/app/views/webui/users/notifications/_notification_filter.html.haml [hunk 5]
```diff
@@ -4,99 +4,99 @@
...
+                                                              checked: selected_filter[:kind]&.include?('reports') }
     .px-4.py-1
-      = render partial: 'webui/shared/check_box', locals: { label: 'Workflow Runs', amount: @count['workflow_runs'],
+      = render partial: 'webui/shared/check_box', locals: { label: 'Workflow Runs', amount: counted_notifications['workflow_runs'],
                                                             key: 'kind[workflow_runs]', name: 'kind[]', value: 'workflow_runs',
                                                             label_icon: 'fas fa-book-open',
-                                                            checked: @selected_filter[:kind]&.include?('workflow_runs') }
-    - if Flipper.enabled?(:content_moderation, @user)
+                                                            checked: selected_filter[:kind]&.include?('workflow_runs') }
+    - if Flipper.enabled?(:content_moderation, user)
```
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/_notification_filter.html.haml:67]:
```suggestion
    - if Flipper.enabled?(:content_moderation, User.session!)
```

File: src/api/app/views/webui/users/notifications/_update.js.erb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-$('#filters').html("<%= j(render NotificationFilterComponent.new(notifications: all_filtered_notifications, selected_filter: selected_filter, user: user)) %>");
+$('#filters').html("<%= j(render partial: 'notification_filter', locals: { selected_filter: selected_filter, counted_notifications: @counted_notifications, user: user }) %>");
```
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/_update.js.erb:1]:
```suggestion
$('#filters').html("<%= j(render partial: 'notification_filter', locals: { selected_filter: selected_filter, counted_notifications: @counted_notifications }) %>");
```

File: src/api/app/views/webui/users/notifications/index.html.haml [hunk 7]
```diff
@@ -10,9 +10,11 @@
...
         %i.float-end.mt-1.fa.fa-chevron-down#notifications-dropdown-trigger
       .collapse#filters
-        = render NotificationFilterComponent.new(notifications: @all_filtered_notifications,
-                                                 selected_filter: @selected_filter,
-                                                 user: User.session)
+        = render partial: 'notification_filter', locals: { selected_filter: @selected_filter,
+                                                           counted_notifications: @counted_notifications,
+                                                           user: User.session,
+                                                           projects_for_filter: @projects_for_filter,
+                                                           groups_for_filter: @groups_for_filter }
```
Comment (hennevogel) [inline src/api/app/views/webui/users/notifications/index.html.haml:17]:
```suggestion
        = render partial: 'notification_filter', locals: { selected_filter: @selected_filter,
                                                           counted_notifications: @counted_notifications,
                                                           projects_for_filter: @projects_for_filter,
                                                           groups_for_filter: @groups_for_filter }
```

**Review by hennevogel — APPROVED** [review]

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 8]
```diff
@@ -67,6 +70,14 @@ def set_notifications
     @notifications = filter_notifications_by_kind(@notifications, @filter_kind)
   end
 
+  def set_notifications_count
+    @counted_notifications = {}
+    @counted_notifications['all'] = @notifications.count
```
Comment (danidoni) [inline src/api/app/controllers/webui/users/notifications_controller.rb:75]: :thought_balloon: Isn't this `@notifications.unread.count + @notifications.read.count`?

Comment (hennevogel) [reply src/api/app/controllers/webui/users/notifications_controller.rb:75]: there is no default state...

Comment (ncounter) [reply src/api/app/controllers/webui/users/notifications_controller.rb:73]: I don't understand what this discussion is about, sorry. Could you explain what do you mean exactly?
Btw yes, `@counted_notifications['all'] = @notifications.count` is the sum of `unread + read` ones.

Comment (hennevogel) [reply src/api/app/controllers/webui/users/notifications_controller.rb:73]: sorry I meant: there is no default state *in the query*

**Review by hennevogel — CHANGES_REQUESTED** [review]
Just read the controller again, I think we count filter notifications only now....

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 9]
```diff
@@ -67,6 +68,14 @@ def set_notifications
...
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:73]: `@notifications` is already filtered by `set_notifications` at this point. So we count only the subset.

Comment (ncounter) [reply src/api/app/controllers/webui/users/notifications_controller.rb:73]: Ah, I see, that's a typo. It was supposed to be `User.notifications.count`. Then I wonder if that would trigger another query. Maybe I'll move the ` @counted_notifications['all'] =` right after `@notifications = User.session!.notifications.for_web.includes(notifiable: [{ commentable: [{ comments: :user }, :project, :bs_request_actions] }, :bs_request_actions, :reviews])` before the filtering part

Comment (ncounter) [reply src/api/app/controllers/webui/users/notifications_controller.rb:79]: https://github.com/openSUSE/open-build-service/pull/16257/commits/41ce5d20227116acd47263a62bf3354cbd2ad22b

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 10]
```diff
@@ -9,6 +9,8 @@ class Webui::Users::NotificationsController < Webui::WebuiController
   before_action :set_filter_kind, :set_filter_state, :set_filter_project, :set_filter_group
   before_action :set_notifications
   before_action :set_notifications_to_be_updated, only: [:update]
+  before_action :set_notifications_count
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:12]:
```suggestion
  before_action :set_counted_notifications
```

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 11]
```diff
@@ -61,6 +72,16 @@ def set_filter_group
 
   def set_notifications
     @notifications = User.session!.notifications.for_web.includes(notifiable: [{ commentable: [{ comments: :user }, :project, :bs_request_actions] }, :bs_request_actions, :reviews])
+  end
+
+  def set_notifications_count
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:77]:
```suggestion
  def set_counted_notifications
```

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 12]
```diff
@@ -9,6 +9,8 @@ class Webui::Users::NotificationsController < Webui::WebuiController
   before_action :set_filter_kind, :set_filter_state, :set_filter_project, :set_filter_group
   before_action :set_notifications
   before_action :set_notifications_to_be_updated, only: [:update]
+  before_action :set_notifications_count
+  before_action :set_filtered_notifications
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:13]:
```suggestion
  before_action :filter_notifications
```

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 13]
```diff
@@ -61,6 +72,16 @@ def set_filter_group
...
+  end
+
+  def set_notifications_count
+    @counted_notifications = {}
+    @counted_notifications['all'] = @notifications.count
+    @counted_notifications['unread'] = @notifications.unread.count
+    @counted_notifications['read'] = @notifications.read.count
+  end
+
+  def set_filtered_notifications
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:84]:
```suggestion
  def filter_notifications
```

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: src/api/app/controllers/webui/users/notifications_controller.rb [hunk 14]
```diff
@@ -24,13 +26,22 @@ def update
     @count = @notifications.where(id: @notification_ids, delivered: !deliver).update_all(delivered: deliver)
     # rubocop:enable Rails/SkipsModelValidations
 
+    # update the count after the update
+    if deliver
+      @counted_notifications['unread'] -= @count
```
Comment (hennevogel) [inline src/api/app/controllers/webui/users/notifications_controller.rb:31]: This will break if either side isn't a number...

Comment (hennevogel) [reply src/api/app/controllers/webui/users/notifications_controller.rb]: pushed a commit to fix this

Comment (ncounter) [reply src/api/app/controllers/webui/users/notifications_controller.rb]: Why do you think `@counted_notifications['unread']` could be anything different than a number?

Anyway, your solution was in my initial approach, but it does not work: if we do the counting after the update like you do now, `@notifications` is already the filtered and paginated subset, so it does not work in this straight forward way. Either we manually adjust the numbers, or we recompute the counting on the entire set, which might slow down the performance again.

Maybe we can do something in a mid-way: count again but only the unread ones (which is expected to be a fast enough query) and deduct the `read` amount by the following:
```
@counted_notifications['unread'] = User.session.unread_notifications
@counted_notifications['read'] = @counted_notifications['all'] - @counted_notifications['unread']
```

Comment (ncounter) [reply src/api/app/controllers/webui/users/notifications_controller.rb]: https://github.com/openSUSE/open-build-service/pull/16257/commits/9fbac3de3ef69a00c49a2f3943dfdf8e9775885c

**Review by hennevogel — APPROVED** [review]

**Pull request merged**